data import: add IMPORT INTO stmt (#13928)
pingcap/docs: sql-statements/sql-statement-import-into.md

---
title: IMPORT INTO
summary: An overview of the usage of IMPORT INTO in TiDB.
---

# IMPORT INTO

> The `IMPORT INTO` statement is used to import data in formats such as `CSV`, `SQL`, `PARQUET` into an empty table in TiDB via the [Physical Import Mode](/tidb-lightning/tidb-lightning-physical-import-mode.md) of TiDB Lightning.

Author: ```suggestion
The `IMPORT INTO` statement is used to import data in formats such as `CSV`, `SQL`, and `PARQUET` into an empty table in TiDB via the [Physical Import Mode](/tidb-lightning/tidb-lightning-physical-import-mode.md) of TiDB Lightning.
```

    - When this framework is enabled ([tidb_enable_dist_task](/system-variables.md#tidb_enable_dist_task-new-in-v710) is `ON`), `IMPORT INTO` distributes subtasks to different TiDB nodes for execution to improve the import efficiency.
    - When this framework is disabled, `IMPORT INTO` only supports running on the TiDB node where the current user is connected.

- For data files stored locally in TiDB, `IMPORT INTO` only supports running on the TiDB node where the current user is connected. Therefore, the data files need to be placed on the TiDB node where the current user is connected.

## Restrictions

- `IMPORT INTO` only supports importing data into existing empty tables in the database.
- `IMPORT INTO` does not support transactions or rollback. Executing `IMPORT INTO` within an explicit transaction (`BEGIN`/`END`) will return an error.
- The execution of `IMPORT INTO` blocks the current connection until the import is completed. To execute the statement asynchronously, you can add the `DETACHED` option.

> - `IMPORT INTO` does not support working simultaneously with features such as TiCDC data replication or [Point-in-Time Recovery (PITR)](/br/br-log-architecture.md).

Author: ```suggestion
- `IMPORT INTO` does not support working simultaneously with features such as [the acceleration of adding indexes](/system-variables.md#tidb_ddl_enable_fast_reorg-new-in-v630), TiDB Lightning data import, TiCDC data replication, or [Point-in-Time Recovery (PITR)](/br/br-log-architecture.md).
```

> - `IMPORT INTO` does not support working simultaneously with features such as [acceleration of adding indexes](/system-variables.md#tidb_ddl_enable_fast_reorg-new-in-v630), data import using TiDB Lightning, data replication using TiCDC, or [Point-in-Time Recovery (PITR)](/br/br-log-architecture.md).

Author: ```suggestion
- `IMPORT INTO` does not support working simultaneously with features such as [FLASHBACK CLUSTER TO TIMESTAMP](/sql-statements/sql-statement-flashback-to-timestamp.md), [acceleration of adding indexes](/system-variables.md#tidb_ddl_enable_fast_reorg-new-in-v630), data import using TiDB Lightning, data replication using TiCDC, or [Point-in-Time Recovery (PITR)](/br/br-log-architecture.md).
```

- Only one `IMPORT INTO` task can run on a cluster at a time. Although `IMPORT INTO` performs a precheck for running tasks, it is not a hard limit. Starting multiple tasks might work when multiple clients execute `IMPORT INTO` simultaneously, but you need to avoid that.

> - Avoid performing DDL and DML operations on the target table during the data import process, because it might cause import failures or data inconsistencies. It is also **NOT** recommended to perform read operations during the import process, as the data being read might be inconsistent. Perform read and write operations only after the import is completed.

Author: ```suggestion
- During the data import process, do not perform DDL or DML operations on the target table, and do not execute [FLASHBACK DATABASE](/sql-statements/sql-statement-flashback-database.md) for the target database. These operations can cause import failures or data inconsistencies. In addition, it is **NOT** recommended to perform read operations during the import process, as the data being read might be inconsistent. Perform read and write operations only after the import is completed.
```

- The import process consumes system resources significantly. To get better performance, it is recommended to use TiDB nodes with at least 32 cores and 64 GiB of memory. TiDB writes sorted data to the TiDB [temporary directory](/tidb-configuration-file.md#temp-dir-new-in-v630) during import, so it is recommended to configure high-performance storage media such as flash memory. For more information, see [Physical Import Mode limitations](/tidb-lightning/tidb-lightning-physical-import-mode.md#requirements-and-restrictions).
- The TiDB [temporary directory](/tidb-configuration-file.md#temp-dir-new-in-v630) is expected to have at least 90 GiB of available space.

## Prerequisites for import

Before using `IMPORT INTO` to import data, make sure the following requirements are met:

- The target table to be imported is already created in TiDB and it is empty.
- The target cluster has sufficient space to store the data to be imported.

> - The [temporary directory](/tidb-configuration-file.md#temp-dir-new-in-v630) of the TiDB node connected to the current session has at least 90 GiB of available space. If [`tidb_enable_dist_task`](/system-variables.md#tidb_enable_dist_task) is enabled, also make sure that the temporary directory of each TiDB node in the cluster has sufficient disk space.

Author: ```suggestion
- The [temporary directory](/tidb-configuration-file.md#temp-dir-new-in-v630) of the TiDB node connected to the current session has at least 90 GiB of available space. If [`tidb_enable_dist_task`](/system-variables.md#tidb_enable_dist_task-new-in-v710) is enabled, also make sure that the temporary directory of each TiDB node in the cluster has sufficient disk space.
```

-->

> **Warning:**
>
> Currently, this statement is experimental. It is not recommended to use it in production environments.

`IMPORT INTO` supports importing data from files stored in Amazon S3, GCS, and the TiDB local storage.

- For data files stored in Amazon S3 or GCS, `IMPORT INTO` supports running in the [TiDB backend task distributed execution framework](/tidb-distributed-execution-framework.md).

    - When this framework is enabled ([tidb_enable_dist_task](/system-variables.md#tidb_enable_dist_task-new-in-v710) is `ON`), `IMPORT INTO` splits a data import job into multiple sub-jobs and distributes these sub-jobs to different TiDB nodes for execution to improve the import efficiency.
    - When this framework is disabled, `IMPORT INTO` only supports running on the TiDB node where the current user is connected.

- For data files stored locally in TiDB, `IMPORT INTO` only supports running on the TiDB node where the current user is connected. Therefore, the data files need to be placed on the TiDB node where the current user is connected. If you access TiDB through a proxy or load balancer, you cannot import data files stored locally in TiDB.

## Restrictions

> 

Author: ```suggestion

Currently, `IMPORT INTO` only supports importing data within 1 TB
```

-->

> **Warning:**
>
> Currently, this statement is experimental. It is not recommended to use it in production environments.

`IMPORT INTO` supports importing data from files stored in Amazon S3, GCS, and the TiDB local storage.

- For data files stored in Amazon S3 or GCS, `IMPORT INTO` supports running in the [TiDB backend task distributed execution framework](/tidb-distributed-execution-framework.md).

    - When this framework is enabled ([tidb_enable_dist_task](/system-variables.md#tidb_enable_dist_task-new-in-v710) is `ON`), `IMPORT INTO` splits a data import job into multiple sub-jobs and distributes these sub-jobs to different TiDB nodes for execution to improve the import efficiency.
    - When this framework is disabled, `IMPORT INTO` only supports running on the TiDB node where the current user is connected.

- For data files stored locally in TiDB, `IMPORT INTO` only supports running on the TiDB node where the current user is connected. Therefore, the data files need to be placed on the TiDB node where the current user is connected. If you access TiDB through a proxy or load balancer, you cannot import data files stored locally in TiDB.

## Restrictions

> Currently, `IMPORT INTO` only supports importing data within 1 TiB.

Author: ```suggestion
- Currently, `IMPORT INTO` only supports importing data within 1 TiB.
```
